File: lua/glacier/widget/taglist.lua
local Posix = require("posix")
local Widget = require("snowcap.widget")

local color = require("glacier.misc.color")

---Glacier's TagList module
---@class glacier.widget.taglist
---@field mt metatable The module metatable
---
---@overload fun(...: glacier.widget.taglist.Config): glacier.widget.TagList
local taglist = { mt = {} }

----------------------
-- Type Definitions --
----------------------

---Render function for Tags.
---@alias glacier.widget.taglist.TagViewCallback fun(tag: glacier.widget.taglist.Tag, style: glacier.widget.taglist.TagStyle): snowcap.widget.WidgetDef?

---Render function for the full list
---@alias glacier.widget.taglist.ListViewCallback fun(tag: snowcap.widget.WidgetDef[], style: glacier.widget.taglist.Style): snowcap.widget.WidgetDef?

---@package
---@class glacier.widget.taglist.ActiveChanged
---@field handle pinnacle.tag.TagHandle
---@field active boolean

---TagList's events
---@class glacier.widget.taglist.Event
---@field toggle? pinnacle.tag.TagHandle
---@field switch? pinnacle.tag.TagHandle
---@field enter? pinnacle.tag.TagHandle
---@field exit? pinnacle.tag.TagHandle
---@field next? {}
---@field prev? {}
---@field small_scroll? {}
---@field active_changed? glacier.widget.taglist.ActiveChanged

---Taglist's messages.
---@class glacier.widget.taglist.Message
---@field id integer
---@field event glacier.widget.taglist.Event
local Message = {}

---Single Tag's appearance.
---@class glacier.widget.taglist.TagStyle
---@field bg_color? snowcap.widget.Color
---@field fg_color? snowcap.widget.Color
---@field font? snowcap.widget.Font
---@field pixels? number
---@field border? snowcap.widget.Border
---@field padding? snowcap.widget.Padding
local TagStyle = {}

---TagList's appearance.
---@class glacier.widget.taglist.Style: glacier.widget.taglist.TagStyle
---@field active? glacier.widget.taglist.TagStyle
---@field inactive? glacier.widget.taglist.TagStyle
---@field spacing? number
---@field hover_transform? fun(style: glacier.widget.taglist.TagStyle): glacier.widget.taglist.TagStyle
local Style = {}

---Single Tag.
---@class glacier.widget.taglist.Tag
---@field name string
---@field handle pinnacle.tag.TagHandle
---@field active boolean
---@field hovered boolean

---TagList's configuration.
---@class glacier.widget.taglist.Config
---@field output pinnacle.output.OutputHandle
---@field style? glacier.widget.taglist.Style
---@field tag_view_callback? glacier.widget.taglist.TagViewCallback
---@field list_view_callback? glacier.widget.taglist.ListViewCallback

---TagList widget.
---
---@class glacier.widget.TagList: snowcap.widget.Program
---@field output pinnacle.output.OutputHandle
---@field tag_view_callback glacier.widget.taglist.TagViewCallback
---@field list_view_callback glacier.widget.taglist.ListViewCallback
---@field private _style glacier.widget.taglist.Style
---@field private _tags glacier.widget.taglist.Tag[]
---@field private _last_scroll number
---@field private _throttle_scroll number
local TagList = setmetatable({}, { __index = require("snowcap.widget.base").Base })

----------------------
-- Module functions --
----------------------

---Default TagList's appearance.
---
---@param style? glacier.widget.taglist.Style
---@return glacier.widget.taglist.Style
function taglist.default_style(style)
    ---@type glacier.widget.taglist.Style
    return require("glacier.utils").merge_table({
        fg_color = color.from_hex("#CCCCCC"),
        font = {
            family = Widget.font.family.Monospace,
            weight = Widget.font.weight.BOLD,
        },
        border = {
            width = 0.0,
        },
        padding = {
            top = 2.0,
            bottom = 2.0,
            left = 8.0,
            right = 8.0,
        },
        spacing = 0.,
        active = {
            bg_color = color.from_hex("#33991A"),
        },
        inactive = {
            bg_color = color.from_hex("#666666"),
        },
    }, style)
end

---Default Tag render function.
---
---@param tag glacier.widget.taglist.Tag
---@param style glacier.widget.taglist.TagStyle
---@return snowcap.widget.WidgetDef?
function taglist.default_tag_view(tag, style)
    return Widget.container({
        child = Widget.text({
            text = tag.name,
            height = Widget.length.Fill,
            width = Widget.length.Shrink,
            valign = Widget.alignment.CENTER,
            style = style:to_text(),
        }),
        padding = style.padding,
        style = style:to_container(),
    })
end

---Default TagList render function.
---
---@param children snowcap.widget.WidgetDef[]
---@param style glacier.widget.taglist.Style
---@return snowcap.widget.WidgetDef?
function taglist.default_list_view(children, style)
    return Widget.row({
        children = children,
        height = Widget.length.Fill,
        width = Widget.length.shrink,
        spacing = style.spacing,
    })
end

---Default hover transform.
---
---@param amount number
---@return fun(style: glacier.widget.taglist.TagStyle): glacier.widget.taglist.TagStyle
function taglist.brighten_background(amount)
    return function(style)
        local clamp = function(value, min, max)
            if value < min then
                return min
            elseif value > max then
                return max
            else
                return value
            end
        end
        local background = style.bg_color

        if background then
            local red = background.red or 0.0
            local green = background.green or 0.0
            local blue = background.blue or 0.0

            background.red = clamp(red * amount, 0.0, 1.0)
            background.green = clamp(green * amount, 0.0, 1.0)
            background.blue = clamp(blue * amount, 0.0, 1.0)
        end

        style.bg_color = background
        return style
    end
end

---@package
---Get all tags tied to an output.
---
---@param output pinnacle.output.OutputHandle
---@return glacier.widget.taglist.Tag[]
local function get_all_tags(output)
    local tag_handles = output:tags()

    local requests = {}
    for i, handle in pairs(tag_handles) do
        requests[i] = function()
            return {
                name = handle:name(),
                active = handle:active(),
            }
        end
    end

    local tags = require("snowcap.util").batch(requests)

    ---@type glacier.widget.taglist.Tag[]
    local tag_list = {}
    for i, tag in pairs(tags) do
        table.insert(tag_list, {
            handle = tag_handles[i],
            name = tag.name,
            active = tag.active,
            hovered = false,
        })
    end

    return tag_list
end

---@package
---
---@return number
local function _now()
    local sec, nsec = Posix.clock_gettime(0)
    return sec + nsec / 10 ^ 9
end

----------------------
-- Message function --
----------------------

---@param id integer
---@param handle pinnacle.tag.TagHandle
---@return glacier.widget.taglist.Message
function Message.toggle(id, handle)
    ---@type glacier.widget.taglist.Message
    return {
        id = id,
        event = {
            toggle = handle,
        },
    }
end

---@param id integer
---@param handle pinnacle.tag.TagHandle
---@return glacier.widget.taglist.Message
function Message.switch(id, handle)
    ---@type glacier.widget.taglist.Message
    return {
        id = id,
        event = {
            switch = handle,
        },
    }
end

---@param id integer
---@param handle pinnacle.tag.TagHandle
---@return glacier.widget.taglist.Message
function Message.enter(id, handle)
    ---@type glacier.widget.taglist.Message
    return {
        id = id,
        event = {
            enter = handle,
        },
    }
end

---@param id integer
---@param handle pinnacle.tag.TagHandle
---@return glacier.widget.taglist.Message
function Message.exit(id, handle)
    ---@type glacier.widget.taglist.Message
    return {
        id = id,
        event = {
            exit = handle,
        },
    }
end

---@param id integer
---@param handle pinnacle.tag.TagHandle
---@return glacier.widget.taglist.Message
function Message.active_changed(id, handle, active)
    ---@type glacier.widget.taglist.Message
    return {
        id = id,
        event = {
            active_changed = {
                handle = handle,
                active = active,
            },
        },
    }
end

----------------------
-- TagStyle methods --
----------------------

---Create a container style from this TagStyle
---@return snowcap.widget.container.Style
function TagStyle:to_container()
    ---@type snowcap.widget.container.Style
    return {
        text_color = self.fg_color,
        background = self.bg_color and Widget.background.Color(self.bg_color) or nil,
        border = self.border,
    }
end

---Create a text style from this TagStyle
function TagStyle:to_text()
    ---@type snowcap.widget.text.Style
    return {
        color = self.fg_color,
        pixels = self.pixels,
        font = self.font,
    }
end

-----------------------
-- TagStyle lifetime --
-----------------------

---@package
---Initialize style
---
---@param style glacier.widget.taglist.TagStyle
---@return glacier.widget.taglist.TagStyle
function TagStyle.new(style)
    style = style or {}

    return setmetatable(style, { __index = TagStyle }) --[[@as glacier.widget.taglist.TagStyle]]
end

-------------------
-- Style methods --
-------------------

---Return a `TagStyle`, based on a tag active state.
---@param active boolean
---@return glacier.widget.taglist.TagStyle
function Style:get(active)
    local deep_copy = require("snowcap.util").deep_copy
    local merge_table = require("glacier.utils").merge_table

    ---@type glacier.widget.taglist.TagStyle
    local style = deep_copy({
        fg_color = self.fg_color,
        bg_color = self.bg_color,
        border = self.border,
        padding = self.padding,
        font = self.font,
    })

    if active then
        style = merge_table(style, self.active)
    else
        style = merge_table(style, self.inactive)
    end

    return TagStyle.new(style)
end

--------------------
-- Style lifetime --
--------------------

---@package
---Initialize style
---
---@param style glacier.widget.taglist.Style
---@return glacier.widget.taglist.Style
function Style.new(style)
    return setmetatable(style, { __index = Style }) --[[@as glacier.widget.taglist.Style]]
end

----------------------------
-- TagList public methods --
----------------------------

-----------------------------
-- TagList private methods --
-----------------------------

---@private
---
---@return snowcap.widget.WidgetDef[]
function TagList:view_tags()
    ---@type snowcap.widget.WidgetDef[]
    local widgets = {}
    local style = self._style
    local view_callback = self.tag_view_callback or taglist.default_tag_view

    for _, tag in pairs(self._tags) do
        local tag_style = style:get(tag.active)
        if tag.hovered and style.hover_transform then
            tag_style = style.hover_transform(tag_style)
        end

        local tag_widget = view_callback(tag, tag_style)
        if tag_widget then
            table.insert(
                widgets,
                Widget.mouse_area({
                    child = tag_widget,
                    on_release = Message.switch(self:id(), tag.handle),
                    on_right_release = Message.toggle(self:id(), tag.handle),
                    on_enter = Message.enter(self:id(), tag.handle),
                    on_exit = Message.exit(self:id(), tag.handle),
                })
            )
        end
    end

    return widgets
end

---@private
---
---@param event snowcap.widget.mouse_area.ScrollEvent
---@return glacier.widget.taglist.Message
function TagList:on_scroll(event)
    local x = 0.0
    local y = 0.0

    if event.lines then
        x = event.lines.x
        y = event.lines.y
    elseif event.pixels then
        x = event.pixels.x
        y = event.pixels.y
    end

    x = x or 0.0
    y = y or 0.0

    local delta = math.abs(x) > math.abs(y) and x or y

    ---@type glacier.widget.taglist.Event
    local msg_event = {}

    if delta > 0.5 then
        msg_event.next = {}
    elseif delta < -0.5 then
        msg_event.prev = {}
    else
        msg_event.small_scroll = {}
    end

    ---@type glacier.widget.taglist.Message
    return {
        id = self:id(),
        event = msg_event,
    }
end

---@private
function TagList:set_hover_for(handle, hover)
    for _, tag in pairs(self._tags) do
        if tag.handle.id == handle.id then
            tag.hovered = hover
            return
        end
    end
end

---@private
---@return integer?
function TagList:find_active_idx()
    for i, tag in ipairs(self._tags) do
        if tag.active then
            return i
        end
    end
end

---@private
function TagList:focus_next_tag()
    local idx = self:find_active_idx()

    if not idx or idx == #self._tags then
        idx = 1
    else
        idx = idx + 1
    end

    self._tags[idx].handle:switch_to()
end

---@private
function TagList:focus_prev_tag()
    local idx = self:find_active_idx()

    if not idx or idx == 1 then
        idx = #self._tags
    else
        idx = idx - 1
    end

    self._tags[idx].handle:switch_to()
end

---@private
---@param msg glacier.widget.taglist.Message|any
---@return glacier.widget.taglist.Event?
function TagList:get_message_event(msg)
    if msg and msg.id == self:id() then
        return msg.event
    end

    return nil
end

---------------------------------
-- impl snowcap.widget.Program --
---------------------------------

---Creates a widget definition for display by Snowcap.
---
---A widget may return nil to notify its parent program that it has
---nothing to display. It's up to the parent to decide whether to display a
---placeholder or to remove the widget from the tree.
---@return snowcap.widget.WidgetDef?
function TagList:view()
    if #self._tags == 0 then
        return
    end

    local children = self:view_tags()

    if not children or #children == 0 then
        return
    end

    local list_view = self.list_view_callback or taglist.default_list_view
    local list = list_view(children, self._style)

    if not list then
        return
    end

    return Widget.mouse_area({
        child = list,
        on_scroll = function(event)
            return self:on_scroll(event)
        end,
    })
end

---Updates this widget program with the received message.
---@param msg any|glacier.widget.prompt.Message
function TagList:update(msg)
    local event = self:get_message_event(msg)

    if not event then
        return
    end

    if event.prev or event.next then
        local now = _now()
        local diff = now - self._last_scroll

        if diff < self._throttle_scroll then
            return
        else
            self._last_scroll = now
        end
    end

    if event.switch then
        event.switch:switch_to()
    elseif event.toggle then
        event.toggle:toggle_active()
    elseif event.enter then
        self:set_hover_for(event.enter, true)
    elseif event.exit then
        self:set_hover_for(event.exit, true)
    elseif event.next then
        self:focus_next_tag()
    elseif event.prev then
        self:focus_prev_tag()
    elseif event.active_changed then
        local handle = event.active_changed.handle
        local active = event.active_changed.active
        for _, tag in pairs(self._tags) do
            if tag.handle.id == handle.id then
                tag.active = active
                break
            end
        end
    end
end

---Called when a surface has been created with this program.
---
---A surface handle is provided to allow the program to manupulate
---the surface. This handle should be passed to any child programs
---to allow them to use it as well.
---
---@param event snowcap.widget.SurfaceEvent
function TagList:event(event)
    if event.created then
        self._tags = get_all_tags(self.output)

        local Tag = require("pinnacle.tag")
        local StdSig = require("snowcap.widget.signal")

        local sig_handle = Tag.connect_signal({
            active = function(handle, active)
                self:emit(StdSig.send_message, Message.active_changed(self:id(), handle, active))
            end,
        })

        self._sig_handle = sig_handle
        self:emit(StdSig.redraw_needed)
    elseif event.closing then
        self._sig_handle:disconnect_all()
    end
end

-----------
-- Other --
-----------

---Create a new TagList widget,
---
---@param config glacier.widget.taglist.Config
---@return glacier.widget.TagList
function TagList:new(config)
    config = config or {}
    config.style = config.style or taglist.default_style()

    local base = require("snowcap.widget.base").Base.new()
    local ret = setmetatable(base, {
        __index = TagList,
    }) --[[@as glacier.widget.TagList]]

    ret.output = config.output
    ret.tag_view_callback = config.tag_view_callback
    ret.list_view_callback = config.list_view_callback

    ret._style = Style.new(config.style)
    ret._tags = {}
    ret._throttle_scroll = 0.05
    ret._last_scroll = 0.0

    return ret
end

---Create a new TagList widget
---
---@param ... glacier.widget.taglist.Config
---@return glacier.widget.TagList
function taglist.mt:__call(...)
    return TagList:new(...)
end

---@diagnostic disable-next-line:param-type-mismatch
return setmetatable(taglist, taglist.mt) --[[@as glacier.widget.taglist]]

File: rust/glacier/src/widget/taglist.rs
//! TagList widget.
//!
//! [`TagList`] allows to display tags, and control them using the mouse. When the view function is
//! called, this widget first render each [`Tag`], wrapping the result in a [`MouseArea`] which
//! will handle left and right click to switch and toggle the given tag, respectively. This first
//! render pass can be overriden through the use of the [`tag_view_callback`] attribute, and
//! default to calling [`TagList::default_tag_view`].
//!
//! The resulting views are then passed to another function with should render the list itself,
//! whose resulting view will be wrapped in a second `MouseArea`, handing Scroll events. This
//! second render pass can be overriden via the [`list_view_callback`] attribute, and default to
//! calling [`TagList::default_list_view`].
//!
//! [`tag_view_callback`]: TagList::tag_view_callback
//! [`list_view_callback`]: TagList::list_view_callback

use std::time::{Duration, Instant};

use pinnacle_api::signal::TagSignal;
use pinnacle_api::util::Batch;
use pinnacle_api::{output::OutputHandle, tag::TagHandle};
use snowcap_api::widget::row::Row;
use snowcap_api::widget::text::Text;
use snowcap_api::widget::{WidgetDef, mouse_area};
use snowcap_api::widget::{container::Container, mouse_area::MouseArea};

use crate::color;
use crate::widget::message::{self, MessageBuilder};
use crate::widget::{WeakState, signal};
use crate::{
    signal::WithEmitter,
    widget::{State, Widget, WidgetMessage, WithState, base::WidgetBase},
};

pub mod style;
#[doc(inline)]
pub use style::{Style, TagStyle, brighten_background};

type TagViewCallback<Msg> = Box<dyn Fn(&Tag, TagStyle) -> Option<WidgetDef<Msg>> + Send + Sync>;
type ListViewCallback<Msg> =
    Box<dyn Fn(Vec<WidgetDef<Msg>>, Style) -> Option<WidgetDef<Msg>> + Send + Sync>;

/// [`TagList`] actions.
#[derive(Clone, Debug)]
pub enum Action {
    Toggle(TagHandle),
    Switch(TagHandle),
    EnterTag(TagHandle),
    ExitTag(TagHandle),
    NextTag,
    PrevTag,
    SmallScroll,
}

/// [`TagList`] messages type.
pub type Message = message::Message<Action>;

/// Single Tag state.
pub struct Tag {
    /// Handle to the tag.
    pub handle: TagHandle,
    /// Name of the tag.
    pub name: String,
    /// Whether the tag is currently active.
    pub active: bool,
    /// Whether the tag [`MouseArea`] is hovered.
    pub hovered: bool,
}

/// Internal [`TagList`] stste.
pub struct Inner<Msg> {
    base: WidgetBase,
    tags: Vec<Tag>,
    style: Style,
    message_builder: MessageBuilder<Action>,
    tag_view_callback: Option<TagViewCallback<Msg>>,
    list_view_callback: Option<ListViewCallback<Msg>>,
    throttle_scroll: Duration,
    last_scroll: Instant,
    _output: OutputHandle,
}

/// Widget representing a list of Tags.
#[derive(Clone)]
pub struct TagList<Msg> {
    state: State<Inner<Msg>>,
}

/// Non-owning version of the [`TagList`].
#[derive(Clone)]
pub struct WeakTagList<Msg>(WeakState<Inner<Msg>>);

/// Default [`TagList`] appearance.
pub fn default_style() -> Style {
    use snowcap_api::widget::{
        Border, Padding,
        font::{self, Font},
    };

    Style::new()
        .fg_color(color::from_hex("#CCCCCC"))
        .font(
            Font::new()
                .family(font::Family::Monospace)
                .weight(font::Weight::Bold),
        )
        .border(Border {
            width: Some(0.0),
            ..Default::default()
        })
        .padding(Padding {
            top: 2.,
            bottom: 2.,
            left: 8.,
            right: 8.,
        })
        .spacing(0.)
        .active(TagStyle::new().bg_color(color::from_hex("#33991A")))
        .inactive(TagStyle::new().bg_color(color::from_hex("#666666")))
}

fn get_all_tags(output: OutputHandle) -> impl Iterator<Item = Tag> {
    output.tags().batch_map(|tag| {
        Box::pin(async {
            Tag::new(
                tag.clone(),
                tag.name_async().await,
                tag.active_async().await,
            )
        })
    })
}

impl<Msg> TagList<Msg>
where
    Msg: Send + 'static,
{
    const DEFAULT_THROTTLE: Duration = Duration::from_millis(50);

    /// Create a new [`TagList`] for a given [`OutputHandle`].
    pub fn new(output: OutputHandle) -> Self {
        let tags = get_all_tags(output.clone()).collect();
        let base = WidgetBase::new("TagList");
        let message_builder = MessageBuilder::new(base.id());

        let state = State::new(Inner {
            base,
            tags,
            style: default_style(),
            message_builder,
            tag_view_callback: None,
            list_view_callback: None,
            throttle_scroll: Self::DEFAULT_THROTTLE,
            last_scroll: Instant::now(),
            _output: output,
        });

        let list = Self { state };
        let weak = list.downgrade();

        pinnacle_api::tag::connect_signal(TagSignal::Active(Box::new(move |handle, active| {
            let Some(list) = weak.upgrade() else {
                return;
            };

            let mut redraw = false;

            {
                let mut inner = list.state.0.lock().unwrap();

                for tag in inner.tags.iter_mut() {
                    if &tag.handle == handle {
                        tag.active = active;
                        redraw = true;
                        break;
                    }
                }
            }

            if redraw {
                list.emit(signal::RedrawNeeded);
            }
        })));

        list
    }

    /// Sets the [`Style`].
    pub fn style(self, style: Style) -> Self {
        self.state.0.lock().unwrap().style = style;
        self.emit(signal::RedrawNeeded);
        self
    }

    /// Sets the callback to render a single [`Tag`].
    pub fn tag_view_callback<F>(self, callback: F) -> Self
    where
        F: Fn(&Tag, TagStyle) -> Option<WidgetDef<Msg>> + Send + Sync + 'static,
    {
        self.state.0.lock().unwrap().tag_view_callback = Some(Box::new(callback));
        self.emit(signal::RedrawNeeded);
        self
    }

    /// Sets the callback to render the full list.
    pub fn list_view_callback<F>(self, callback: F) -> Self
    where
        F: Fn(Vec<WidgetDef<Msg>>, Style) -> Option<WidgetDef<Msg>> + Send + Sync + 'static,
    {
        self.state.0.lock().unwrap().list_view_callback = Some(Box::new(callback));
        self.emit(signal::RedrawNeeded);
        self
    }

    /// Sets the cooldown time between two Scroll events.
    pub fn throttle_scroll(self, throttle: Duration) -> Self {
        self.state.0.lock().unwrap().throttle_scroll = throttle;
        self
    }

    /// Create a new [`WeakTagList`] for this `TagList`.
    pub fn downgrade(&self) -> WeakTagList<Msg> {
        WeakTagList(self.state.downgrade())
    }

    /// Default view to render [`Tag`].
    pub fn default_tag_view(tag: &Tag, style: TagStyle) -> Option<WidgetDef<Msg>> {
        let text = Text::new(tag.name.clone())
            .height(snowcap_api::widget::Length::Fill)
            .width(snowcap_api::widget::Length::Shrink)
            .vertical_alignment(snowcap_api::widget::Alignment::Center)
            .style(style.clone().into());

        let mut widget = Container::new(text).style(style.clone().into());
        widget.padding = style.padding;

        Some(widget.into())
    }

    /// Default view to render a list of [`Tag`].
    pub fn default_list_view(
        children: Vec<WidgetDef<Msg>>,
        style: Style,
    ) -> Option<WidgetDef<Msg>> {
        let mut widget = Row::new_with_children(children)
            .height(snowcap_api::widget::Length::Fill)
            .width(snowcap_api::widget::Length::Shrink);

        widget.spacing = style.spacing;

        Some(widget.into())
    }
}

impl<Msg> WeakTagList<Msg> {
    /// Attempts to upgrade this `WeakTagList` to a [`TagList`].
    ///
    /// Returns [`None`] if the [`TagList`] has already been dropped.
    pub fn upgrade(&self) -> Option<TagList<Msg>> {
        self.0.upgrade().map(|state| TagList { state })
    }
}

impl<Msg> Inner<Msg>
where
    Msg: From<WidgetMessage> + Send + 'static,
{
    fn view_tags(&self) -> Vec<WidgetDef<Msg>> {
        self.tags
            .iter()
            .filter_map(|t| {
                let style = self.style.get(t.active);
                let style = if t.hovered
                    && let Some(transform) = &self.style.hover_transform
                {
                    transform(style)
                } else {
                    style
                };

                let builder = self.message_builder;

                let tag = if let Some(callback) = &self.tag_view_callback {
                    callback(t, style)
                } else {
                    TagList::default_tag_view(t, style)
                };

                tag.map(|c| {
                    MouseArea::new(c)
                        .on_release(builder.switch(t.handle.clone()).into())
                        .on_right_release(builder.toggle(t.handle.clone()).into())
                        .on_enter(builder.enter(t.handle.clone()).into())
                        .on_exit(builder.exit(t.handle.clone()).into())
                        .into()
                })
            })
            .collect()
    }

    fn set_hover_for(&mut self, handle: TagHandle, hover: bool) {
        for t in self.tags.iter_mut() {
            if t.handle == handle {
                t.hovered = hover;
                return;
            }
        }
    }

    fn find_active_idx(&self) -> Option<usize> {
        self.tags
            .iter()
            .enumerate()
            .find(|(_, t)| t.active)
            .map(|(idx, _)| idx)
    }

    fn focus_next_tag(&mut self) {
        if let Some(idx) = self.find_active_idx() {
            let idx = (idx + 1) % self.tags.len();

            self.tags[idx].handle.switch_to();
        } else if let Some(t) = self.tags.first_mut() {
            t.handle.switch_to();
        }
    }

    fn focus_prev_tag(&mut self) {
        if let Some(idx) = self.find_active_idx() {
            let idx = if idx == 0 { self.tags.len() } else { idx };
            let idx = idx - 1;

            self.tags[idx].handle.switch_to();
        } else if let Some(t) = self.tags.last_mut() {
            t.handle.switch_to();
        }
    }

    fn on_scroll(builder: MessageBuilder<Action>, delta: mouse_area::ScrollDelta) -> Msg {
        let delta = match delta {
            mouse_area::ScrollDelta::Lines { x, y } => {
                if f32::abs(x) > f32::abs(y) {
                    x
                } else {
                    y
                }
            }
            mouse_area::ScrollDelta::Pixels { x, y } => {
                if f32::abs(x) > f32::abs(y) {
                    x
                } else {
                    y
                }
            }
        };

        if delta > 0.5 {
            builder.next_tag().into()
        } else if delta < -0.5 {
            builder.prev_tag().into()
        } else {
            builder.small_scroll().into()
        }
    }
}

impl<Msg> WithEmitter for TagList<Msg> {
    fn with_emitter(&self) -> crate::signal::Emitter {
        self.state.0.lock().unwrap().with_emitter()
    }
}

impl<Msg> WithState for TagList<Msg> {
    type Type = Inner<Msg>;

    fn with_state(&self) -> State<Self::Type> {
        self.state.clone()
    }
}

impl<Msg> WithEmitter for Inner<Msg> {
    fn with_emitter(&self) -> crate::signal::Emitter {
        self.base.with_emitter()
    }
}

impl<Msg> Widget for Inner<Msg>
where
    Msg: Clone + From<WidgetMessage> + Into<Option<WidgetMessage>> + Send + Sync + 'static,
{
    type Message = Msg;

    fn view(&self) -> Option<snowcap_api::widget::WidgetDef<Self::Message>> {
        if self.tags.is_empty() {
            return None;
        };

        let style = self.style.clone();
        let children = self.view_tags();
        let builder = self.message_builder;

        let list = if let Some(callback) = &self.list_view_callback {
            callback(children, style)
        } else {
            TagList::default_list_view(children, style)
        }?;

        let widget = MouseArea::new(list).on_scroll(move |delta| Self::on_scroll(builder, delta));

        Some(widget.into())
    }

    fn update(&mut self, msg: Self::Message) {
        let Some(msg) = msg.into() else {
            return;
        };

        let action = match msg {
            WidgetMessage::TagList(Message { id, action }) if id == self.base.id() => action,
            _ => return,
        };

        if matches!(action, Action::NextTag | Action::PrevTag) {
            let now = Instant::now();
            let diff = now - self.last_scroll;

            if diff < self.throttle_scroll {
                return;
            } else {
                self.last_scroll = now;
            }
        }

        match action {
            Action::Switch(t) => t.switch_to(),
            Action::Toggle(t) => t.toggle_active(),
            Action::EnterTag(t) => self.set_hover_for(t, true),
            Action::ExitTag(t) => self.set_hover_for(t, false),
            Action::NextTag => self.focus_next_tag(),
            Action::PrevTag => self.focus_prev_tag(),
            _ => {}
        }
    }
}

impl Tag {
    fn new(handle: TagHandle, name: String, active: bool) -> Self {
        Self {
            handle,
            name,
            active,
            hovered: false,
        }
    }
}

impl From<Message> for WidgetMessage {
    fn from(value: Message) -> Self {
        Self::TagList(value)
    }
}

impl MessageBuilder<Action> {
    fn toggle(&self, handle: TagHandle) -> WidgetMessage {
        self.build(Action::Toggle(handle))
    }

    fn switch(&self, handle: TagHandle) -> WidgetMessage {
        self.build(Action::Switch(handle))
    }

    fn enter(&self, handle: TagHandle) -> WidgetMessage {
        self.build(Action::EnterTag(handle))
    }

    fn exit(&self, handle: TagHandle) -> WidgetMessage {
        self.build(Action::ExitTag(handle))
    }

    fn next_tag(&self) -> WidgetMessage {
        self.build(Action::NextTag)
    }

    fn prev_tag(&self) -> WidgetMessage {
        self.build(Action::PrevTag)
    }

    fn small_scroll(&self) -> WidgetMessage {
        self.build(Action::SmallScroll)
    }
}
